### PR TITLE
[SPARK-27989] [Core] Added retries on the connection to the driver for k8s

### PR DESCRIPTION
**Pull request by jlpedrosa**
Disabled negative dns caching for docker images

Improved logging on DNS resolution, convenient for slow k8s clusters

## What changes were proposed in this pull request?
Added retries when building the connection to the driver in K8s.
In some scenarios DNS reslution can take more than the timeout.
Also openjdk-8 by default has negative dns caching enabled, which means even retries may not help depending on the times.

## How was this patch tested?
This patch was tested agains an specific k8s cluster with slow response time in DNS to ensure it woks.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (felixcheung) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: this changes not just for k8s, right?

Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @felixcheung 
It's the executor used for kubernetes but It canbe called for different resource managers.
I Labeled it as K8s because it also modifies the docker images used in k8s. Not sure what would be the best component. For instance YarnCoarseGrainedExecutorBackend extends this class.

Comment (felixcheung) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: yes, that's what I mean. in case of shared class, the PR should tag all of them

Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @felixcheung 
Then this patch belongs to core (due to the retries) and Kubernetes due to the java property in the docker image. I've modified the title as such.

Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @felixcheung  any feedback?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (skonto) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: shouldnt this be configurable?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (skonto) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: It would be more flexible if that could be set along with `[networkaddress.cache.ttl]`(https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/java-dg-jvm-ttl.html) via an env variable from within the entrypoint script.

File: core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala
Comment (skonto) [inline core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: remove space here and bellow, pls follow the format the other tests have.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (skonto) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: Do you have a sample output of this log entry?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Nit, but a new method tailrec is big overkill here. Just
```for (_ <- 0 until 3) { ... }```

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (jlpedrosa) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: It is a boolean, so "... isUnresolved true|false"

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: As I read in the docs, no those values can't be overriden via properties. It can be just replaced the whole file. 
https://stackoverflow.com/questions/4521119/java-argument-to-specify-java-security-file-for-jvm
You can use this code to test.

```
import java.security.Security;

public class HelloWorld {

    public static void main(String[] args) {
       String prop = Security.getProperty("networkaddress.cache.negative.ttl");
       System.out.println("java.security.networkaddress.cache.negative.ttl: " + prop);
    }
}
```

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @srowen  
This helps the code to be testable, if you want I can change it. But I'd also have to delete the tests. I'll wait for your feedback.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (srowen) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: Nit: to be clearer this could be `isUnresolved=true` etc. Or could get fancier and make the message change to either "DNS resolution failed for ..." or "DNS resolution succeeded for .."

File: core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala
Comment (jlpedrosa) [reply core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: Hi @skonto 

This class has a Main method and reads the values from the argv. In my opinon is an overkill to add it to define the numbers of the retries. Let me know if what you want is to me add arguments to the main method or you prefer me to leave it as it is.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @skonto

This class has a Main method and reads the values from the argv. In my opinon is an overkill to add it to define the numbers of the retries. Let me know if what you want is to me add arguments to the main method or you prefer me to leave it as it is.

File: core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala
Comment (srowen) [reply core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: I just don't think this test or method is worthwhile. It's more or less verifying you've simulated a for loop.

Comment (jlpedrosa) [reply core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: It tests mainly two things: 1) That you indeed go n times, and 2) the behaviour of throwing the exception.   I can change it if you want, for sure there is not performance difference and I think it will be actually less lines of code than the try catch and the if statements. 
@srowen let me know if you want me to go to the iterative approach.

Comment (srowen) [reply core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: I would keep it simple and just write a for loop.

Comment (jlpedrosa) [reply core/src/test/scala/org/apache/spark/executor/CoarseGrainedExecutorBackendSuite.scala]: Test will be deleted then.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @srowen 
I've done as you requested and transformed it into a for loop.

**Comment by jlpedrosa** [comment]
Please note that I've squashed the commits and rebased it.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (skonto) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: My thought was that someone may want to increase this value because it depends on the env.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (skonto) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: I still find it better to be able to load your own file: `java.security.properties`. I dont agree with this being hardcode.

**Comment by skonto** [comment]
@jlpedrosa pls relate your PR with a jira ticket.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: `0 until nTries`

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: The try-catch indent is off, but, this doesn't do anything? `i` can't reach `nTries`.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (srowen) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: I still think this log statement can be slightly improved if we're bothering.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (srowen) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: I don't think we disable negative resolution caching anywhere. It has a potential negative impact, though, don't know if on the whole it's right. I'd leave it out of this otherwise minor change.

But, isn't the property just `networkaddress.cache.negative.ttl`? 
https://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html
At least that should be settable as a simple system property. Not sure if it's indeed different when trying to go this route.

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @srowen true, I'll fix it.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: True.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Yes, I understand, my question still remains. This is part of a Main methof. Do you want me to add it to the args read or not...?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Hi @srowen @skonto 
The problem with Java and negative resolution is the moment you have ANY timeout in DNS resolution, you'll never be able to resolve that name. Which in some Kubernetes scenarios happens (pods network taking time to get fully operational). Basically no matter how many retries it will never work if you have a timeout.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (jlpedrosa) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: Fixed in code.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: @skonto  About the options you said, I saw the option of sending the whole file,  I thought it was too complicated for people, and I think this issue is mostly present in K8s.

@srowen && @skonto  how do we continue with this?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: I personally would leave this. Not worth and arg or config yet.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: I think this can still just be simpler.
```
var driver: RpcEndpointRef = null
for (i <- 0 until 3) {
  try {
    driver = ...
  } catch {
    case e: Throwable => // continue
  }
}
require(driver != null, "...")
```

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (srowen) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: The default seems to be networkaddress.cache.negative.ttl = 10 or 10 seconds. Never caching might be better, but we don't do this anywhere else and seems like something that can be left to the user. Yeah OK it isn't a different property you're talking about and does seem only settable in the security props file, which isn't easy.
Never caching or caching for 10 seconds probably won't affect whether 1 or 3 tries fails or succeeds here, if that's the issue.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: @srowen  I agree. @skonto  you ok? or how do i move forward, not sure what is the "consensus" for this type of cases.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Yes @srowen ,  exactly, it's 10 seconds. The issue comes with the actual look up which is a native invocation in jdk8, which has a timeout completely different to those 10 seconds, managed by the os.
So actually the code is not retrying at all because the results are cached. Unless the amount of retries would be absurdly high or we start sleeping in the code. In my opinion a retry limit of 10 seconds is way excessive. If you think about it the OS already provides this functionality.

Comment (srowen) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: So, if we disable negative caching in the JVM, you mean the OS will still cache the negative resolution? does that leave the same problem, or do I misunderstand? are you just saying this is an attempt to get rid of the negative caching and may not succeed depending on OS?
Well, I'm not super against it if the K8S folks are OK with it, as this part is specific to K8S, but then, diverges from how it's run elsewhere, potentially.

Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Hi @srowen 
Let me answer properly. AFAIK that DNS negative lookup is NOT cached, not at least in linux, positive caching yes, retries and timeouts, and it also depends on distribution, not all of them have it enabled (and there are different ways to achieve so).

What I meant is due to the fact that is cached, then retries won't work because they'll happen in a tight loop (the for loop we were adding), unless we start adding sleeps in the scala code, which may or may not be enough depending on the resolution timeouts at OS level. The end result of the caching of the negative resolution at java code is that final users end up with a very strange behaviour where the timeouts of the OS level are only respected once. 

What will happen if we don't disable caching, the first call to resolve, will wait the OS timeout (I think 5 seconds is the default), then the subsequent calls will just don't even try because the java layer won't try to invoke it. [here](https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/net/InetAddress.java#L1251) and [here](https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/net/InetAddress.java#L885)

So what this means is, if we don't disable negative DNS caches, then retries won't be able to solve the issue it in a reliable manner (a complicated combination of retries, sleeps in scala, and OS timers would be able to solve it).

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Actually If we don't add the conditions int  the for loop, it will try to get many connections, even if they are succeeding, so I fixed the other way.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (skonto) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: @jlpedrosa
>  I saw the option of sending the whole file, I thought it was too complicated for people, and I think this issue is mostly present in K8s.
I am not saying to send the whole the file, that is one option, user could just create it on the fly in his entrypoint script based on env vars. In my custom image for example I have a Prometheus config file that works for all my deployments, users in a slow env may even choose to have these security properties predefined in the file in the image. I dont see why you need to hardcoded them. 
The dockerfile you are modifying is about k8s only so we care about k8s here only.

Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Hi @skonto 
For me any option is ok, the entrypoint for the container, I'll adapt to whatever you guys prefer. I am aware that the Dockerfile is only used in k8s, the train of thoughts is that this scenario is only common at k8s when pods networking may take some time (CNI, Sidecar pods..). Just let me know what is the path you preffer and I'll implement in such way.
In the docker image that spark uses (the base [spark image](https://github.com/apache/spark/blob/1217996f1574f758d8cccc1c4e3846452d24b35b/resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile#L18) is openjdk8 alpine [openjdk8](https://github.com/docker-library/openjdk/blob/178c542fbb93a8f8a42e331b73a1214c9d8ba81d/8/jdk/alpine/Dockerfile#L7))  dns caching is disabled, if you'd start an openjdk8 image and your try:
`watch -n 2 ping -c 1  www.apache.org`
while tcpdumping, you'll see that every 2 seconds a request is issued.
I tried this also in an ubtuntu docker image, and I get the same behaviour.

@skonto @srowen: Just let me know the way you'd prefer it and I'll patch it that way.

Comment (skonto) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Hi @jlpedrosa I voted for making it configurable and extendable ;) We should also make a note of this eg. docs? Let's see what @srowen has to say and has the final call.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Oops @jlpedrosa I meant to retain the condition in the loop, but otherwise I find the code right above simpler. However I like retaining the original exception. Just please pay attention to the style. `driver = ` indent is off and needs to say `if (i == nTries - 1)`

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (srowen) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: I just wouldn't make this change here.

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: @srowen 
Fixed the indentation.  Please double check.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Hi @srowen,
Ok, then you'd you be ok to do the approach proposed by @skonto and have it as config option in the entrypoint script for the docker image?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: It looks the same; did you push?

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (srowen) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: I'm neutral on that. It seems like a non-trivial change to the whole JVM to work around an env-specific DNS issue.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: @srowen  I think it's in the right place, I'm rewriting history to clean the multiple commits. I think it is in the right place...

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile]: Hi @srowen @skonto 

I've added the option as you considered it was acceptable. Please have a look and let me know what you think. This at least would enable to set it up via spark operator, but I can open a different PR if you want that I modify the code to manage that env variable in the scala side as a managed option for 2.4. In 3.0 it can be handled via pod templates, have not tested it, but I think all scala is doing append variables, so defining that variable in a template could enable it

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (srowen) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Nit: -1 => - 1 (I think this might fail the style checker)

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (srowen) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: The indent is too deep; can it just do on the previous line? if not continuation indent is at most 4 spaces

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh
Comment (srowen) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh]: Again, I just wouldn't do this here. It's pretty unrelated. I am neutral on adding another flag for this. You're really working around a pretty specific failure mode here on the JVM side.

Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh]: @srowen @skonto 
I am afraid I am not following you. @skonto pointed out that he preferred the fix instead on Docker file, in the entrypoint.sh script with a configurable variable (which is also specific for k8s) so this is what this patch has. As you said "I wouldn't do this here" I though you also meant the Dockerfile and you agreed.

So I am not sure what do you mean by "wouldn't do this here", where is here and where is the  place where you want it?

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (jlpedrosa) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: Hi @srowen 
I thought the line length was limited too. Pushed the fix.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala
Comment (jlpedrosa) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Hi @srowen Ah, Ok. Didn't saw that space, if there's a way to run the style checker would be excelent! I followed the [official docs](http://spark.apache.org/contributing.html#code-style-guide) but that tried to refactor hundreds of files. I've pushed the change you requested.

Comment (srowen) [reply core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Run `./dev/lint-scala` to just run the checks

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh
Comment (srowen) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh]: I mean that a) I'm not sure we should make this change but if we do, it's b) not directly related to the JIRA/PR purpose, which is just to introduce retries. I think you're saying there is not much point in retrying unless this change happens too?

Comment (jlpedrosa) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh]: hi @srowen 

Now I understand what you mean, "here" means this PR or jira ticket (please correct me if I miss understood you). Then no worries, I'll keep the retries only in this one and then start a new PR/Jira process to discuss the DNS, and see what and where we can do.
In my opinion retries make sense, avoid rescheduling something for a transient error that is recoverable is helpful, this is not multi layered system so it can't storm out.

I've re-written history into a single commit to make it cleaner.

JL

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master. (The 'does not merge' message is spurious.)

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java
Comment (srowen) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java]: Arg, the line is too long and fails Java style checks. I'll fix. I don't see why the last test run didn't catch it ... the 'does not merge' message isn't related.

**Pull request closed without merging**
